src/invariant/ArchivedStateConsistency.cpp

```diff
@@ -0,0 +1,519 @@
+
+    while (archivedIt != archived.end() && liveIt != live.end())
+    {
+        if (archivedIt->first < liveIt->first)
+        {
+            archivedIt++;
+            continue;
+        }
+        else if (liveIt->first < archivedIt->first)
+        {
+            liveIt++;
+            continue;
+        }
+        else
+        {
+            return fmt::format(
+                FMT_STRING(
+                    "ArchivedStateConsistency:: Entry with the same key is "
+                    "present in both live and archived state. Key: {}"),
+                xdrToCerealString(archivedIt->first, "entry_key"));
+        }
+    }
+
+    CLOG_INFO(Invariant, "ArchivedStateConsistency invariant passed");
+    return std::string{};
+}
+
+std::shared_ptr<Invariant>
+ArchivedStateConsistency::registerInvariant(Application& app)
+{
+    return app.getInvariantManager()
+        .registerInvariant<ArchivedStateConsistency>();
+}
+
+std::string
+ArchivedStateConsistency::getName() const
+{
+    return "ArchivedStateConsistency";
+}
+
+std::string
+ArchivedStateConsistency::checkOnLedgerCommit(
+    SearchableSnapshotConstPtr lclLiveState,
+    SearchableHotArchiveSnapshotConstPtr lclHotArchiveState,
+    std::vector<LedgerEntry> const& evictedFromLive,
+    std::vector<LedgerKey> const& deletedKeysFromLive,
+    UnorderedMap<LedgerKey, LedgerEntry> const& restoredFromArchive,
+    UnorderedMap<LedgerKey, LedgerEntry> const& restoredFromLiveState)
+{
+    LogSlowExecution logSlow("ArchivedStateConsistency::checkOnLedgerCommit",
+                             LogSlowExecution::Mode::AUTOMATIC_RAII, "took",
+                             std::chrono::milliseconds(1));
+
+    if (protocolVersionIsBefore(
+            lclLiveState->getLedgerHeader().ledgerVersion,
+            LiveBucket::FIRST_PROTOCOL_SUPPORTING_PERSISTENT_EVICTION))
+    {
+        CLOG_INFO(Invariant,
+                  "Skipping ArchivedStateConsistency invariant for "
+                  "protocol version {}",
+                  lclLiveState->getLedgerHeader().ledgerVersion);
+        return std::string{};
+    }
+    auto ledgerSeq = lclLiveState->getLedgerSeq() + 1;
+    auto ledgerVers = lclLiveState->getLedgerHeader().ledgerVersion;
+
+    // Collect all keys to preload
+    LedgerKeySet allKeys;
+
+    // Keys for evicted from live entries
+    for (auto const& e : evictedFromLive)
+    {
+        auto key = LedgerEntryKey(e);
+        allKeys.insert(key);
+        if (isPersistentEntry(key))
+        {
+            allKeys.insert(getTTLKey(e));
+        }
+    }
+
+    // Keys for deleted from live (temp and TTLs)
+    for (auto const& k : deletedKeysFromLive)
+    {
+        allKeys.insert(k);
+        if (isPersistentEntry(k))
+        {
+            allKeys.insert(getTTLKey(k));
+        }
+    }
+
+    // Keys for restored entries
+    for (auto const& [key, entry] : restoredFromArchive)
+    {
+        allKeys.insert(key);
+        if (isPersistentEntry(key))
```

Comment:
Both restoredFromArchive and restoredFromLiveState should asset all keys are persistent and remove the branch.

Reply:
Bump, I think this would still be a helpful assert